proposals/0000-updated-partial-type-sigs.rst

```diff
@@ -0,0 +1,726 @@
+   the existence of an underscore in a visible kind application has GHC treat a type signature as partial,
+   even though GHC does not stop compilation or emit warnings for such underscores. Furthermore, the value
+   of an underscore in a visible kind application (e.g., in ``Proxy @_ True``) can often be inferred from
+   context, not from an expression. This design was chosen because it's close to the treatment for visible
+   type applications, but it's an unhappy compromise.
+
+   Visible kind application (even an innocuous usage like the example in the previous paragraph) with underscores is not allowed
+   where wildcards are not allowed, such as in data constructor declarations and in instance heads.
+
+   This bit is really just an outright bug, but it's unclear how to fix this bug without a proposal such
+   as this one.
+
+3. Named wildcards act like wildcards but are named. This allows two niceties: the user can specify that
+   the same wildcard is used twice, and output is clarified by giving a name to the wildcard (instead of
+   just ``_``). Here is an example::
+
+     foo :: _w -> _w -> _w
+     foo x 'z' = x
+
+   GHC will infer that ``foo :: Char -> Char -> Char`` (and that ``_w`` stands
+   for ``Char``), knowing that the second argument must be a ``Char`` and that
+   both arguments and the return type must be the same. This feature is
+   enabled with ``-XNamedWildCards``.
+
+   The feature is undiscoverable. If I write the code above in a module without ``-XNamedWildCards``, I get a type
+   error. This is because ``_w`` is a legal type variable name in standard Haskell. No error message in GHC suggests enabling this
+   extension.
+
+   Naming is useful for all form of underscore except for ignoreds, as knowing that several different
+   underscores mean the same thing can aid inference.
+
+4. The current design of partial type signatures treats type generalization and constraint generalization differently.
+   Consider these examples::
+
+     ex1 :: _ -> _
+     ex1 x = x
+
+     ex2 :: _ -> _
+     ex2 x = x + 1
+
+     ex3 :: _ -> _
+     ex3 x = not x
+
+   With ``-XPartialTypeSignatures`` enabled, ``ex1`` and ``ex3`` are accepted, but ``ex2`` is rejected. (Actually, ``ex2``
+   is accepted because of type defaulting. Say ``default ()`` to disable type defaulting, and you will observe that it
+   is rejected. Avoiding this twist in the narrative would complicate the example unnecessarily.) In ``ex1``, GHC discovers
+   that ``_`` stands for an unconstrained type variable ``t``, generalizes, and gets ``ex1 :: t -> t``. In ``ex3``, GHC
+   discovers that ``_`` standards for ``Bool``. In ``ex2``, GHC discovers that ``_`` stands for ``t`` where ``Num t`` must
+   hold; however, it rejects the declaration because no context was specified. In order to accept ``ex2``, we need to
+   write ::
+
+     ex2 :: _ => _ -> _
```

Comment:
This seems like a plain bug to me. A type wildcard should imply a constraint wildcard for any type variables in the inferred type (but perhaps not if those tyvars appear in specified parts of the type).